test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/CaseSensitiveTagHelperDescriptorComparer.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.Razor.TagHelpers;
+using Microsoft.Internal.Web.Utils;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class CaseSensitiveTagHelperDescriptorComparer : TagHelperDescriptorComparer, IEqualityComparer<TagHelperDescriptor>
+    {
+        public new static readonly CaseSensitiveTagHelperDescriptorComparer Default =
+            new CaseSensitiveTagHelperDescriptorComparer();
+
+        private CaseSensitiveTagHelperDescriptorComparer()
+        {
+        }
+
+        bool IEqualityComparer<TagHelperDescriptor>.Equals(TagHelperDescriptor descriptorX, TagHelperDescriptor descriptorY)
+        {
+            return base.Equals(descriptorX, descriptorY) &&
+                // Normal comparer doesn't care about case, in tests we do.
+                string.Equals(descriptorX.TagName, descriptorY.TagName, StringComparison.Ordinal) &&
+                string.Equals(descriptorX.Prefix, descriptorY.Prefix, StringComparison.Ordinal) &&
+                Enumerable.SequenceEqual(
+                    descriptorX.RequiredAttributes,
+                    descriptorY.RequiredAttributes,
+                    StringComparer.OrdinalIgnoreCase) &&
+                descriptorX.Attributes.SequenceEqual(
+                    descriptorY.Attributes,
+                    CaseSensitiveAttributeDescriptorComparer.Default);
+        }
+
+        int IEqualityComparer<TagHelperDescriptor>.GetHashCode(TagHelperDescriptor descriptor)
+        {
+            var hashCodeCombiner = HashCodeCombiner
+                .Start()
+                .Add(base.GetHashCode())
+                .Add(descriptor.TagName, StringComparer.Ordinal)
+                .Add(descriptor.Prefix)
+                .Add(descriptor.Attributes);
```

Comment:
including `Attributes` should mean the hash code is always unique; not what you want and the collection is covered just below